### PR TITLE
 Add autocomplete for /giveachievement

### PR DESCRIPTION
**Pull request by VishramKidPG123**
self-explanatory (look at my other pull request)
also added a thing in the giveachievement function where previously, thanksforplaying could not be removed with commands, but now it can

### CODE REVIEW

**Comment by milena-kos** [comment]
why does this pr also has the other one in it :staring_cat:

**Comment by VishramKidPG123** [comment]
> why does this pr also has the other one in it :staring_cat:

uhh I forked it from the other one
I can do a new pull request without those if you want

**Comment by frinkifail** [comment]
can this be closed

File: main.py [hunk 1]
```diff
@@ -2518,12 +2518,8 @@ async def forcespawn(message: discord.Interaction, cat_type: Optional[str]):
 @discord.app_commands.default_permissions(manage_guild=True)
 @discord.app_commands.rename(person_id="user", ach_id="name")
 @discord.app_commands.describe(person_id="who", ach_id="name or id of the achievement")
-<<<<<<< HEAD
 @discord.app_commands.autocomplete(ach_id=ach_autocomplete)
-async def giveachievement(message: discord.Interaction, person_id: discord.Member, ach_id: str):
-=======
 async def giveachievement(message: discord.Interaction, person_id: discord.User, ach_id: str):
```
Comment (frinkifail) [inline main.py:2522]: user -> member

**Review by frinkifail — CHANGES_REQUESTED** [review]

File: main.py (hunk 1 above)
Comment (VishramKidPG123) [reply main.py:2522]: why? isn't that the opposite of what we're doing?

**Comment by VishramKidPG123** [comment]
> can this be closed

why
I just updated

**Comment by milena-kos** [comment]
or im gonna hit you with a comically big spoon

**Comment by erzahoq** [comment]
you should do it anyway I think

**Comment by VishramKidPG123** [comment]
should ach_id validity checking be removed or should it be kept as a failsafe

**Comment by milena-kos** [comment]
not only as failsafe, autocompletes are just suggestions

**Comment by frinkifail** [comment]
> frinkifail stop sabotaging

im not :<

**Comment by VishramKidPG123** [comment]
any issues?

**Comment by milena-kos** [comment]
okay but why is this straight up not working :staring_cat:

**Comment by VishramKidPG123** [comment]
idk

**Comment by frinkifail** [comment]
what

**Comment by milena-kos** [comment]
like the autocomplete is just straight up empty

**Comment by frinkifail** [comment]
idk man sounds like a skill issue

**Comment by aar0n-py** [comment]
> idk man sounds like a skill issue

yes

**Comment by VishramKidPG123** [comment]
> 
> 
> 
> 
> > idk man sounds like a skill issue
> 
> yes

that was so helpful of you

**Comment by Vest1gemusic** [comment]
if you cat bot if you cat bot if you cat bot if you cat bot

**Comment by milena-kos** [comment]
after fixing the error logs i now finally know why

File: main.py [hunk 2]
```diff
@@ -358,6 +358,14 @@ async def achemb(message, ach_id, send_type, author_string=None):
 async def cat_type_autocomplete(interaction: discord.Interaction, current: str) -> list[discord.app_commands.Choice[str]]:
     return [discord.app_commands.Choice(name=choice, value=choice) for choice in cattypes if current.lower() in choice.lower()][:25]
 
+# converts string to lowercase alphanumeric characters only
+def alnum(string):
+    return "".join(item for item in string.lower() if item.isalnum())
+
+# function to autocomplete achievement choice for /giveachievement, which also allows more than 25 options
+async def ach_autocomplete(interaction: discord.Interaction, current: str) -> list[discord.app_commands.Choice[str]]:
+    return [discord.app_commands.Choice(name=val, value=val) for (key, val) in ach_list.items() if (alnum(current) in alnum(key) or alnum(current) in alphanumeric(val))][:25]
```
Comment (milena-kos) [inline main.py:367]: val here is the entire `{'title': 'First cat!', 'description': 'The journey begins...', 'is_hidden': False, 'category': 'Cat Hunt'}` dict

**Comment by milena-kos** [comment]
ig .title works

**Comment by milena-kos** [comment]
god dangit
image

**Pull request merged**